File: lua/copilot/session.lua
--- CopilotSession: represents a single conversation session with the Copilot CLI.
--
-- Sessions maintain conversation state, handle events, and manage tool execution.
-- Sessions are created via CopilotClient:create_session() or resumed via
-- CopilotClient:resume_session().
--
-- Usage:
--   local session = client:create_session({ model = "gpt-4" })
--   session:on(function(event) print(event.type) end)
--   session:send({ prompt = "Hello!" })
--   session:destroy()

local cjson = require("cjson")
local types = require("copilot.types")

-- ---------------------------------------------------------------------------
-- CopilotSession class
-- ---------------------------------------------------------------------------

local CopilotSession = {}
CopilotSession.__index = CopilotSession

--- Create a new CopilotSession (internal, called by CopilotClient).
-- @param session_id string      The unique session ID
-- @param rpc_client JsonRpcClient  The JSON-RPC client
-- @param workspace_path string|nil  The workspace path for infinite sessions
-- @return CopilotSession
function CopilotSession.new(session_id, rpc_client, workspace_path)
    local self = setmetatable({}, CopilotSession)

    self.session_id        = session_id
    self.workspace_path    = workspace_path or ""
    self._rpc_client       = rpc_client
    self._handlers         = {}           -- array of {id=number, fn=function}
    self._next_handler_id  = 1
    self._tool_handlers    = {}           -- toolName -> handler function
    self._permission_handler = nil        -- function or nil
    self._user_input_handler = nil        -- function or nil
    self._hooks            = nil          -- SessionHooks table or nil
    self._exit_plan_mode_handler = nil    -- function or nil
    self._trace_context_provider = nil    -- function or nil

    return self
end

-- ---------------------------------------------------------------------------
-- Public API
-- ---------------------------------------------------------------------------

--- Send a message to this session.
-- @param options table  MessageOptions with fields: prompt, attachments, mode, responseFormat, imageOptions
-- @return string|nil messageId
-- @return string|nil error
function CopilotSession:send(options)
    options = options or {}
    local req = {
        sessionId      = self.session_id,
        prompt         = options.prompt or "",
        attachments    = options.attachments,
        mode           = options.mode,
        responseFormat = options.responseFormat,
        imageOptions   = options.imageOptions,
    }

    -- Inject trace context if provider is available
    if self._trace_context_provider then
        local tc_ok, tc = pcall(self._trace_context_provider)
        if tc_ok and tc then
            if tc.traceparent then req.traceparent = tc.traceparent end
            if tc.tracestate then req.tracestate = tc.tracestate end
        end
    end

    local result, err = self._rpc_client:request("session.send", req)
    if err then
        return nil, "failed to send message: " .. tostring(err)
    end

    if result and result.messageId then
        return result.messageId, nil
    end
    return nil, nil
end

--- Send a message and wait until the session becomes idle.
-- Returns the final assistant message event, or nil if none was received.
-- @param options table     MessageOptions
-- @param timeout number|nil  Timeout in seconds (default 60)
-- @return table|nil  The last assistant.message SessionEvent
-- @return string|nil error
function CopilotSession:send_and_wait(options, timeout)
    timeout = timeout or 60

    local last_assistant_message = nil
    local idle_received = false
    local error_received = nil

    -- Subscribe to events
    local unsubscribe = self:on(function(event)
        if event.type == types.SessionEventType.ASSISTANT_MESSAGE then
            last_assistant_message = event
        elseif event.type == types.SessionEventType.SESSION_IDLE then
            idle_received = true
        elseif event.type == types.SessionEventType.SESSION_ERROR then
            local msg = "session error"
            if event.data and event.data.message then
                msg = event.data.message
            end
            error_received = msg
        end
    end)

    -- Send the message
    local _, send_err = self:send(options)
    if send_err then
        unsubscribe()
        return nil, send_err
    end

    -- Wait for idle or error, polling the read loop
    local start_time = os.time()
    while not idle_received and not error_received do
        -- Check timeout
        if os.time() - start_time >= timeout then
            unsubscribe()
            return nil, "timeout waiting for session.idle"
        end

        -- Drive the read loop for one message
        if self._rpc_client and self._rpc_client._read_one_message then
            local ok, read_err = self._rpc_client:_read_one_message()
            if not ok then
                unsubscribe()
                return nil, "read error: " .. tostring(read_err)
            end
        else
            -- Fallback: small sleep to avoid busy-waiting
            -- In a real async environment, this would yield to the event loop
            local sleep_until = os.clock() + 0.01
            while os.clock() < sleep_until do end
        end
    end

    unsubscribe()

    if error_received then
        return nil, "session error: " .. tostring(error_received)
    end

    return last_assistant_message, nil
end

--- Subscribe to events from this session.
-- Returns an unsubscribe function.
-- @param handler function(event)  Event handler callback
-- @return function  Call to unsubscribe
function CopilotSession:on(handler)
    local id = self._next_handler_id
    self._next_handler_id = self._next_handler_id + 1

    table.insert(self._handlers, { id = id, fn = handler })

    -- Return unsubscribe function
    return function()
        for i, h in ipairs(self._handlers) do
            if h.id == id then
                table.remove(self._handlers, i)
                break
            end
        end
    end
end

--- Retrieve all events and messages from this session's history.
-- @return table|nil  Array of SessionEvent
-- @return string|nil error
function CopilotSession:get_messages()
    local result, err = self._rpc_client:request("session.getMessages", {
        sessionId = self.session_id,
    })
    if err then
        return nil, "failed to get messages: " .. tostring(err)
    end

    if result and result.events then
        return result.events, nil
    end
    return {}, nil
end

--- Retrieve metadata for this session.
-- @return table|nil  The session metadata
-- @return string|nil error
function CopilotSession:get_metadata()
    local result, err = self._rpc_client:request("session.getMetadata", {
        sessionId = self.session_id,
    })
    if err then
        return nil, "failed to get metadata: " .. tostring(err)
    end

    return result, nil
end

--- Destroy this session and release resources.
-- @return boolean success
-- @return string|nil error
function CopilotSession:destroy()
    local _, err = self._rpc_client:request("session.destroy", {
        sessionId = self.session_id,
    })
    if err then
        return false, "failed to destroy session: " .. tostring(err)
    end

    -- Clear all handlers
    self._handlers           = {}
    self._tool_handlers      = {}
    self._permission_handler = nil
    self._user_input_handler = nil
    self._hooks              = nil
    self._exit_plan_mode_handler = nil
    self._trace_context_provider = nil

    return true, nil
end

--- Abort the currently processing message in this session.
-- @return boolean success
-- @return string|nil error
function CopilotSession:abort()
    local _, err = self._rpc_client:request("session.abort", {
        sessionId = self.session_id,
    })
    if err then
        return false, "failed to abort session: " .. tostring(err)
    end
    return true, nil
end

-- ---------------------------------------------------------------------------
-- Internal: event dispatch
-- ---------------------------------------------------------------------------

--- Dispatch an event to all registered handlers.
-- @param event table  SessionEvent
function CopilotSession:_dispatch_event(event)
    if not event then return end

    -- Copy handler list to avoid issues if handlers unsubscribe during iteration
    local handlers = {}
    for _, h in ipairs(self._handlers) do
        table.insert(handlers, h.fn)
    end

    for _, fn in ipairs(handlers) do
        local ok, err = pcall(fn, event)
        if not ok then
            io.stderr:write(string.format(
                "Error in session event handler: %s\n", tostring(err)
            ))
        end
    end
end

-- ---------------------------------------------------------------------------
-- Internal: tool registration
-- ---------------------------------------------------------------------------

--- Register tool handlers for this session.
-- @param tools table|nil  Array of Tool tables (with handler field)
function CopilotSession:_register_tools(tools)
    self._tool_handlers = {}
    if not tools then return end

    for _, tool in ipairs(tools) do
        if tool.name and tool.name ~= "" and tool.handler then
            self._tool_handlers[tool.name] = tool.handler
        end
    end
end

--- Get a registered tool handler by name.
-- @param name string
-- @return function|nil
function CopilotSession:_get_tool_handler(name)
    return self._tool_handlers[name]
end

-- ---------------------------------------------------------------------------
-- Internal: permission handler
-- ---------------------------------------------------------------------------

--- Register a permission handler.
-- @param handler function(request, invocation) -> PermissionRequestResult
function CopilotSession:_register_permission_handler(handler)
    self._permission_handler = handler
end

--- Get the registered permission handler.
-- @return function|nil
function CopilotSession:_get_permission_handler()
    return self._permission_handler
end

-- ---------------------------------------------------------------------------
-- Internal: user input handler
-- ---------------------------------------------------------------------------

--- Register a user input handler.
-- @param handler function(request, invocation) -> UserInputResponse
function CopilotSession:_register_user_input_handler(handler)
    self._user_input_handler = handler
end

--- Get the registered user input handler.
-- @return function|nil
function CopilotSession:_get_user_input_handler()
    return self._user_input_handler
end

-- ---------------------------------------------------------------------------
-- Internal: hooks
-- ---------------------------------------------------------------------------

--- Register hooks for this session.
-- @param hooks table  SessionHooks
function CopilotSession:_register_hooks(hooks)
    self._hooks = hooks
end

--- Get the registered hooks.
-- @return table|nil
function CopilotSession:_get_hooks()
    return self._hooks
end

-- ---------------------------------------------------------------------------
-- Internal: exit plan mode handler
-- ---------------------------------------------------------------------------

--- Register an exit plan mode handler.
-- @param handler function(request) -> ExitPlanModeResponse
function CopilotSession:_register_exit_plan_mode_handler(handler)
    self._exit_plan_mode_handler = handler
end

--- Get the registered exit plan mode handler.
-- @return function|nil
function CopilotSession:_get_exit_plan_mode_handler()
    return self._exit_plan_mode_handler
end

-- ---------------------------------------------------------------------------
-- Internal: trace context provider
-- ---------------------------------------------------------------------------

--- Register a trace context provider.
-- @param provider function() -> TraceContext
function CopilotSession:_register_trace_context_provider(provider)
    self._trace_context_provider = provider
end

--- Handle a hooks invocation from the server.
-- @param hook_type string    The hook type (e.g. "preToolUse", "postToolUse")
-- @param input table         The hook input data
-- @return table|nil output
-- @return string|nil error
function CopilotSession:_handle_hooks_invoke(hook_type, input)
    local hooks = self._hooks
    if not hooks then
        return nil, nil
    end

    local invocation = { sessionId = self.session_id }

    if hook_type == "preToolUse" then
        if not hooks.onPreToolUse then return nil, nil end
        local ok, result = pcall(hooks.onPreToolUse, input, invocation)
        if not ok then return nil, tostring(result) end
        return result, nil

    elseif hook_type == "postToolUse" then
        if not hooks.onPostToolUse then return nil, nil end
        local ok, result = pcall(hooks.onPostToolUse, input, invocation)
        if not ok then return nil, tostring(result) end
        return result, nil

    elseif hook_type == "userPromptSubmitted" then
        if not hooks.onUserPromptSubmitted then return nil, nil end
        local ok, result = pcall(hooks.onUserPromptSubmitted, input, invocation)
        if not ok then return nil, tostring(result) end
        return result, nil

    elseif hook_type == "sessionStart" then
        if not hooks.onSessionStart then return nil, nil end
        local ok, result = pcall(hooks.onSessionStart, input, invocation)
        if not ok then return nil, tostring(result) end
        return result, nil

    elseif hook_type == "sessionEnd" then
        if not hooks.onSessionEnd then return nil, nil end
        local ok, result = pcall(hooks.onSessionEnd, input, invocation)
        if not ok then return nil, tostring(result) end
        return result, nil

    elseif hook_type == "errorOccurred" then
        if not hooks.onErrorOccurred then return nil, nil end
        local ok, result = pcall(hooks.onErrorOccurred, input, invocation)
        if not ok then return nil, tostring(result) end
        return result, nil

    else
        return nil, "unknown hook type: " .. tostring(hook_type)
    end
end

return CopilotSession

File: ruby/lib/copilot/session.rb
# frozen_string_literal: true

# Copyright (c) Microsoft Corporation. All rights reserved.

require "thread"

module Copilot
  # Represents a single conversation session with the Copilot CLI.
  #
  # A session maintains conversation state, handles events, and manages tool execution.
  # Sessions are created via {CopilotClient#create_session} or resumed via
  # {CopilotClient#resume_session}.
  #
  # @example Basic usage
  #   session = client.create_session(model: "gpt-4")
  #
  #   # Subscribe to all events
  #   unsub = session.on { |event| puts event.type }
  #
  #   # Send a message and wait for completion
  #   response = session.send_and_wait(prompt: "Hello!")
  #   puts response&.data&.dig("content")
  #
  #   session.destroy
  class CopilotSession
    # @return [String] the unique session identifier
    attr_reader :session_id

    # @return [String, nil] workspace path when infinite sessions are enabled
    attr_reader :workspace_path

    # @api private
    # @param session_id     [String]
    # @param rpc_client     [JsonRpcClient]
    # @param workspace_path [String, nil]
    def initialize(session_id, rpc_client, workspace_path = nil)
      @session_id     = session_id
      @rpc_client     = rpc_client
      @workspace_path = workspace_path

      @event_handlers      = []
      @typed_event_handlers = {} # type => [handler]
      @event_handlers_lock = Mutex.new

      @tool_handlers      = {}
      @tool_handlers_lock = Mutex.new

      @permission_handler      = nil
      @permission_handler_lock = Mutex.new

      @user_input_handler      = nil
      @user_input_handler_lock = Mutex.new

      @hooks      = nil
      @hooks_lock = Mutex.new

      @exit_plan_mode_handler      = nil
      @exit_plan_mode_handler_lock = Mutex.new

      @trace_context_provider = nil
    end

    # Send a message to this session.
    #
    # @param prompt          [String]            the message text
    # @param attachments     [Array, nil]        optional file/directory/selection attachments
    # @param mode            [String, nil]       "enqueue" (default) or "immediate"
    # @param response_format [String, nil]       desired response format ("text", "image", "json_object")
    # @param image_options   [ImageOptions, nil] options for image generation
    # @return [String] the message ID
    def send(prompt:, attachments: nil, mode: nil, response_format: nil, image_options: nil)
      payload = { sessionId: @session_id, prompt: prompt }
      payload[:attachments] = attachments if attachments
      payload[:mode] = mode if mode
      payload[:responseFormat] = response_format if response_format
      payload[:imageOptions] = image_options.to_h if image_options

      # Inject trace context if provider is available
      if @trace_context_provider
        begin
          tc = @trace_context_provider.call
          if tc
            tp = tc.respond_to?(:traceparent) ? tc.traceparent : tc[:traceparent]
            ts = tc.respond_to?(:tracestate) ? tc.tracestate : tc[:tracestate]
            payload[:traceparent] = tp if tp
            payload[:tracestate] = ts if ts
          end
        rescue StandardError
          # ignore trace context errors
        end
      end

      response = @rpc_client.request("session.send", payload)
      response["messageId"]
    end

    # Send a message and wait until the session becomes idle.
    #
    # This is a convenience method that combines {#send} with waiting for
    # the +session.idle+ event.
    #
    # @param prompt          [String]            the message text
    # @param attachments     [Array, nil]        optional attachments
    # @param mode            [String, nil]       delivery mode
    # @param response_format [String, nil]       desired response format ("text", "image", "json_object")
    # @param image_options   [ImageOptions, nil] options for image generation
    # @param timeout         [Numeric]           seconds to wait (default 60)
    # @return [SessionEvent, nil] the final assistant.message event, or nil
    # @raise [Timeout::Error] if the timeout expires before session.idle
    # @raise [RuntimeError] if the session emits a session.error event
    def send_and_wait(prompt:, attachments: nil, mode: nil, response_format: nil, image_options: nil, timeout: 60)
      idle_mutex = Mutex.new
      idle_cv    = ConditionVariable.new
      idle_fired = false

      last_assistant_message = nil
      error_event = nil

      # Register BEFORE send to avoid race condition
      unsub = on do |event|
        case event.type
        when SessionEventType::ASSISTANT_MESSAGE
          last_assistant_message = event
        when SessionEventType::SESSION_IDLE
          idle_mutex.synchronize do
            idle_fired = true
            idle_cv.signal
          end
        when SessionEventType::SESSION_ERROR
          error_event = RuntimeError.new(
            event.data.is_a?(Hash) ? event.data["message"] : event.data.to_s
          )
          idle_mutex.synchronize do
            idle_fired = true
            idle_cv.signal
          end
        end
      end

      begin
        self.send(prompt: prompt, attachments: attachments, mode: mode,
                  response_format: response_format, image_options: image_options)

        idle_mutex.synchronize do
          unless idle_fired
            idle_cv.wait(idle_mutex, timeout)
          end
        end

        raise error_event if error_event

        unless idle_fired
          raise Timeout::Error, "Timeout after #{timeout}s waiting for session.idle"
        end

        last_assistant_message
      ensure
        unsub.call
      end
    end

    # Subscribe to events from this session.
    #
    # When called with a block only, subscribes to all events.
    # When called with an event type and a block, subscribes to that specific type.
    #
    # @overload on(&handler)
    #   @yield [event] called for every session event
    #   @yieldparam event [SessionEvent]
    #   @return [Proc] unsubscribe function
    #
    # @overload on(event_type, &handler)
    #   @param event_type [String] the specific event type to listen for
    #   @yield [event] called only for events matching the type
    #   @yieldparam event [SessionEvent]
    #   @return [Proc] unsubscribe function
    def on(event_type = nil, &handler)
      raise ArgumentError, "Block required" unless handler

      @event_handlers_lock.synchronize do
        if event_type
          (@typed_event_handlers[event_type] ||= []) << handler
        else
          @event_handlers << handler
        end
      end

      -> {
        @event_handlers_lock.synchronize do
          if event_type
            @typed_event_handlers[event_type]&.delete(handler)
          else
            @event_handlers.delete(handler)
          end
        end
      }
    end

    # Retrieve all events/messages from this session's history.
    #
    # @return [Array<SessionEvent>]
    def get_messages
      response = @rpc_client.request("session.getMessages", { sessionId: @session_id })
      events = response["events"] || []
      events.map { |e| SessionEvent.from_hash(e) }
    end

    # Retrieve metadata for this session.
    #
    # @return [Hash] the session metadata
    def get_metadata
      @rpc_client.request("session.getMetadata", { sessionId: @session_id })
    end

    # Destroy this session and release associated resources.
    def destroy
      @rpc_client.request("session.destroy", { sessionId: @session_id })
      @event_handlers_lock.synchronize do
        @event_handlers.clear
        @typed_event_handlers.clear
      end
      @tool_handlers_lock.synchronize { @tool_handlers.clear }
      @permission_handler_lock.synchronize { @permission_handler = nil }
      @user_input_handler_lock.synchronize { @user_input_handler = nil }
      @hooks_lock.synchronize { @hooks = nil }
      @exit_plan_mode_handler_lock.synchronize { @exit_plan_mode_handler = nil }
      @trace_context_provider = nil
    end

    # Abort the currently processing message.
    def abort
      @rpc_client.request("session.abort", { sessionId: @session_id })
    end

    # ---- Internal API (used by CopilotClient) ----

    # @api private
    def _dispatch_event(event)
      handlers = @event_handlers_lock.synchronize do
        typed = @typed_event_handlers[event.type]&.dup || []
        wildcard = @event_handlers.dup
        typed + wildcard
      end

      handlers.each do |handler|
        handler.call(event)
      rescue StandardError => e
        $stderr.puts("[CopilotSDK] Session event handler error: #{e.class}: #{e.message}")
      end
    end

    # @api private
    def _register_tools(tools)
      @tool_handlers_lock.synchronize do
        @tool_handlers.clear
        return unless tools

        tools.each do |tool|
          next unless tool.name && tool.handler

          @tool_handlers[tool.name] = tool.handler
        end
      end
    end

    # @api private
    def _get_tool_handler(name)
      @tool_handlers_lock.synchronize { @tool_handlers[name] }
    end

    # @api private
    def _register_permission_handler(handler)
      @permission_handler_lock.synchronize { @permission_handler = handler }
    end

    # @api private
    def _handle_permission_request(request)
      handler = @permission_handler_lock.synchronize { @permission_handler }
      unless handler
        return { kind: PermissionKind::DENIED_NO_APPROVAL }
      end

      begin
        perm_request = PermissionRequest.from_hash(request)
        result = handler.call(perm_request, { session_id: @session_id })
        result.is_a?(PermissionRequestResult) ? result.to_h : result
      rescue StandardError
        { kind: PermissionKind::DENIED_NO_APPROVAL }
      end
    end

    # @api private
    def _register_user_input_handler(handler)
      @user_input_handler_lock.synchronize { @user_input_handler = handler }
    end

    # @api private
    def _handle_user_input_request(params)
      handler = @user_input_handler_lock.synchronize { @user_input_handler }
      raise "User input requested but no handler registered" unless handler

      request = UserInputRequest.from_hash(params)
      result = handler.call(request, { session_id: @session_id })
      result.is_a?(UserInputResponse) ? result.to_h : result
    end

    # @api private
    def _register_hooks(hooks)
      @hooks_lock.synchronize { @hooks = hooks }
    end

    # @api private
    def _register_exit_plan_mode_handler(handler)
      @exit_plan_mode_handler_lock.synchronize { @exit_plan_mode_handler = handler }
    end

    # @api private
    def _handle_exit_plan_mode_request(params)
      handler = @exit_plan_mode_handler_lock.synchronize { @exit_plan_mode_handler }
      unless handler
        return { approved: true }
      end

      begin
        request = ExitPlanModeRequest.from_hash(params)
        result = handler.call(request)
        result.is_a?(ExitPlanModeResponse) ? result.to_h : result
      rescue StandardError
        { approved: true }
      end
    end

    # @api private
    def _register_trace_context_provider(provider)
      @trace_context_provider = provider
    end

    # @api private
    def _handle_hooks_invoke(hook_type, input_data)
      hooks = @hooks_lock.synchronize { @hooks }
      return nil unless hooks

      handler_map = {
        "preToolUse"          => hooks.on_pre_tool_use,
        "postToolUse"         => hooks.on_post_tool_use,
        "userPromptSubmitted" => hooks.on_user_prompt_submitted,
        "sessionStart"        => hooks.on_session_start,
        "sessionEnd"          => hooks.on_session_end,
        "errorOccurred"       => hooks.on_error_occurred,
      }

      handler = handler_map[hook_type]
      return nil unless handler

      begin
        handler.call(input_data, { session_id: @session_id })
      rescue StandardError
        nil
      end
    end
  end
end
